# Sales Analysis

### Import libraries

In [3]:
import pandas as pd
import os

In [6]:
files = list(os.listdir('./sales Data'))
print(*files, sep = '\n')

Sales_December_2019.csv
Sales_April_2019.csv
Sales_February_2019.csv
Sales_March_2019.csv
Sales_August_2019.csv
Sales_May_2019.csv
Sales_November_2019.csv
Sales_October_2019.csv
Sales_January_2019.csv
Sales_September_2019.csv
Sales_July_2019.csv
Sales_June_2019.csv


as we have 12 months of separeted data, first we gonna merge all of those files into single dataset

In [7]:
# but before doing that, let's see what the data looks like
pd.read_csv('./sales Data/Sales_April_2019.csv')

Order ID                     Product Quantity Ordered Price Each  \
0       176558        USB-C Charging Cable                2      11.95   
1          NaN                         NaN              NaN        NaN   
2       176559  Bose SoundSport Headphones                1      99.99   
3       176560                Google Phone                1        600   
4       176560            Wired Headphones                1      11.99   
...        ...                         ...              ...        ...   
18378   194090                Google Phone                1        600   
18379   194091       AA Batteries (4-pack)                1       3.84   
18380   194092      AAA Batteries (4-pack)                2       2.99   
18381   194093       AA Batteries (4-pack)                1       3.84   
18382   194094    Lightning Charging Cable                1      14.95   

           Order Date                        Purchase Address  
0      04/19/19 08:46            917 1st St, Dallas, TX 75001  
1                 NaN                                     NaN  
2      04/07/19 22:30       682 Chestnut St, Boston, MA 02215  
3      04/12/19 14:38    669 Spruce St, Los Angeles, CA 90001  
4      04/12/19 14:38    669 Spruce St, Los Angeles, CA 90001  
...               ...                                     ...  
18378  04/08/19 17:11   177 Jackson St, Los Angeles, CA 90001  
18379  04/15/19 16:02         311 Forest St, Austin, TX 73301  
18380  04/28/19 14:36  347 Sunset St, San Francisco, CA 94016  
18381  04/14/19 15:09         835 Lake St, Portland, OR 97035  
18382  04/18/19 11:08          354 North St, Boston, MA 02215  

[18383 rows x 6 columns]

so, this is how our data looks like. it has 6 columns including the time and date of purchase of a specific product.  
let's merge the data

In [8]:
all_df = pd.concat([pd.read_csv(f'./sales data/{file}') for file in files], ignore_index = True)
all_df

Order ID                 Product Quantity Ordered Price Each  \
0        295665      Macbook Pro Laptop                1       1700   
1        295666      LG Washing Machine                1      600.0   
2        295667    USB-C Charging Cable                1      11.95   
3        295668        27in FHD Monitor                1     149.99   
4        295669    USB-C Charging Cable                1      11.95   
...         ...                     ...              ...        ...   
186845   222905  AAA Batteries (4-pack)                1       2.99   
186846   222906        27in FHD Monitor                1     149.99   
186847   222907    USB-C Charging Cable                1      11.95   
186848   222908    USB-C Charging Cable                1      11.95   
186849   222909  AAA Batteries (4-pack)                1       2.99   

            Order Date                        Purchase Address  
0       12/30/19 00:01  136 Church St, New York City, NY 10001  
1       12/29/19 07:03     562 2nd St, New York City, NY 10001  
2       12/12/19 18:21    277 Main St, New York City, NY 10001  
3       12/22/19 15:13     410 6th St, San Francisco, CA 94016  
4       12/18/19 12:38           43 Hill St, Atlanta, GA 30301  
...                ...                                     ...  
186845  06/07/19 19:02           795 Pine St, Boston, MA 02215  
186846  06/01/19 19:29   495 North St, New York City, NY 10001  
186847  06/22/19 18:57   319 Ridge St, San Francisco, CA 94016  
186848  06/26/19 18:35    916 Main St, San Francisco, CA 94016  
186849  06/25/19 14:33          209 11th St, Atlanta, GA 30301  

[186850 rows x 6 columns]

Concatenation of all the data is done. let's save that complete data into a single file

In [9]:
all_df.to_csv('all_data.csv', index = False)

let's read the updated dataframe

In [11]:
data = pd.read_csv('all_data.csv')
data

Order ID                 Product Quantity Ordered Price Each  \
0        295665      Macbook Pro Laptop                1       1700   
1        295666      LG Washing Machine                1      600.0   
2        295667    USB-C Charging Cable                1      11.95   
3        295668        27in FHD Monitor                1     149.99   
4        295669    USB-C Charging Cable                1      11.95   
...         ...                     ...              ...        ...   
186845   222905  AAA Batteries (4-pack)                1       2.99   
186846   222906        27in FHD Monitor                1     149.99   
186847   222907    USB-C Charging Cable                1      11.95   
186848   222908    USB-C Charging Cable                1      11.95   
186849   222909  AAA Batteries (4-pack)                1       2.99   

            Order Date                        Purchase Address  
0       12/30/19 00:01  136 Church St, New York City, NY 10001  
1       12/29/19 07:03     562 2nd St, New York City, NY 10001  
2       12/12/19 18:21    277 Main St, New York City, NY 10001  
3       12/22/19 15:13     410 6th St, San Francisco, CA 94016  
4       12/18/19 12:38           43 Hill St, Atlanta, GA 30301  
...                ...                                     ...  
186845  06/07/19 19:02           795 Pine St, Boston, MA 02215  
186846  06/01/19 19:29   495 North St, New York City, NY 10001  
186847  06/22/19 18:57   319 Ridge St, San Francisco, CA 94016  
186848  06/26/19 18:35    916 Main St, San Francisco, CA 94016  
186849  06/25/19 14:33          209 11th St, Atlanta, GA 30301  

[186850 rows x 6 columns]

In [14]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 186850 entries, 0 to 186849
Data columns (total 6 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   Order ID          186305 non-null  object
 1   Product           186305 non-null  object
 2   Quantity Ordered  186305 non-null  object
 3   Price Each        186305 non-null  object
 4   Order Date        186305 non-null  object
 5   Purchase Address  186305 non-null  object
dtypes: object(6)
memory usage: 8.6+ MB


as we observe that, Dtype of every column is object. it means that the data present is either string or mixed type.  
for more information see [https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.dtypes.html](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.dtypes.html)

let's search for null values in the data

In [16]:
data.isnull().sum()

Order ID            545
Product             545
Quantity Ordered    545
Price Each          545
Order Date          545
Purchase Address    545
dtype: int64

so we have 545 null values and we gonna drop'em all

In [17]:
data.dropna(inplace = True)
data.isnull().sum()

Order ID            0
Product             0
Quantity Ordered    0
Price Each          0
Order Date          0
Purchase Address    0
dtype: int64

## Data augmentation
let's add month column to the data

In [23]:
data['month'] = data['Order Date'].str[:2].astype('int32')

ValueError: invalid literal for int() with base 10: 'Or'

the reason why we're doing string slicing is that, the format of the date is **mm/dd/yyyy**. so if we grab the first 2 chars from that, we can get the job done

we have encountered an error. it says that it cannot convert string to int. the string is 'Or'.  
so, let's remove all the columns that has 'Or' in the month place of the order date.

In [28]:
data.loc[data['Order Date'].str[:2] == 'Or'] 

Order ID  Product  Quantity Ordered  Price Each  Order Date  \
254     Order ID  Product  Quantity Ordered  Price Each  Order Date   
705     Order ID  Product  Quantity Ordered  Price Each  Order Date   
1101    Order ID  Product  Quantity Ordered  Price Each  Order Date   
2875    Order ID  Product  Quantity Ordered  Price Each  Order Date   
3708    Order ID  Product  Quantity Ordered  Price Each  Order Date   
...          ...      ...               ...         ...         ...   
183671  Order ID  Product  Quantity Ordered  Price Each  Order Date   
184012  Order ID  Product  Quantity Ordered  Price Each  Order Date   
184041  Order ID  Product  Quantity Ordered  Price Each  Order Date   
184275  Order ID  Product  Quantity Ordered  Price Each  Order Date   
186532  Order ID  Product  Quantity Ordered  Price Each  Order Date   

        Purchase Address  
254     Purchase Address  
705     Purchase Address  
1101    Purchase Address  
2875    Purchase Address  
3708    Purchase Address  
...                  ...  
183671  Purchase Address  
184012  Purchase Address  
184041  Purchase Address  
184275  Purchase Address  
186532  Purchase Address  

[355 rows x 6 columns]

as we can see, some of the rows don't have proper date. instead they have string 'Order Date' as their data. 

In [32]:
# or we can also do this instead to get the same result but in a different way
# let's try for searching for the most repeated value in Order Date column. so obviously it's gonna be 'Order Date'
data['Order Date'].value_counts()

Order Date        355
12/15/19 20:16      8
12/11/19 13:24      7
10/30/19 21:28      7
04/02/19 13:24      7
                 ... 
08/06/19 08:32      1
08/18/19 20:27      1
08/30/19 16:29      1
08/15/19 14:05      1
06/25/19 14:33      1
Name: Order Date, Length: 142396, dtype: int64

so as we can see, Order Date value has repeated 355 times. so we can remove every row in the data which has Order Date in its Order Date Column

In [33]:
data.drop(data[data['Order Date'] == 'Order Date'].index, inplace = True)

In [34]:
data['Order Date'].value_counts()

12/15/19 20:16    8
10/30/19 21:28    7
04/02/19 13:24    7
12/11/19 13:24    7
10/11/19 19:44    6
                 ..
08/06/19 08:32    1
08/18/19 20:27    1
08/30/19 16:29    1
08/15/19 14:05    1
06/25/19 14:33    1
Name: Order Date, Length: 142395, dtype: int64

now we don't have the Order Date value anymore in the dataset

In [35]:
# adding month column to the data
data['month'] = data['Order Date'].str[:2].astype('int32')
data

Order ID                 Product Quantity Ordered Price Each  \
0        295665      Macbook Pro Laptop                1       1700   
1        295666      LG Washing Machine                1      600.0   
2        295667    USB-C Charging Cable                1      11.95   
3        295668        27in FHD Monitor                1     149.99   
4        295669    USB-C Charging Cable                1      11.95   
...         ...                     ...              ...        ...   
186845   222905  AAA Batteries (4-pack)                1       2.99   
186846   222906        27in FHD Monitor                1     149.99   
186847   222907    USB-C Charging Cable                1      11.95   
186848   222908    USB-C Charging Cable                1      11.95   
186849   222909  AAA Batteries (4-pack)                1       2.99   

            Order Date                        Purchase Address  month  
0       12/30/19 00:01  136 Church St, New York City, NY 10001     12  
1       12/29/19 07:03     562 2nd St, New York City, NY 10001     12  
2       12/12/19 18:21    277 Main St, New York City, NY 10001     12  
3       12/22/19 15:13     410 6th St, San Francisco, CA 94016     12  
4       12/18/19 12:38           43 Hill St, Atlanta, GA 30301     12  
...                ...                                     ...    ...  
186845  06/07/19 19:02           795 Pine St, Boston, MA 02215      6  
186846  06/01/19 19:29   495 North St, New York City, NY 10001      6  
186847  06/22/19 18:57   319 Ridge St, San Francisco, CA 94016      6  
186848  06/26/19 18:35    916 Main St, San Francisco, CA 94016      6  
186849  06/25/19 14:33          209 11th St, Atlanta, GA 30301      6  

[185950 rows x 7 columns]

In [36]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 185950 entries, 0 to 186849
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   Order ID          185950 non-null  object
 1   Product           185950 non-null  object
 2   Quantity Ordered  185950 non-null  object
 3   Price Each        185950 non-null  object
 4   Order Date        185950 non-null  object
 5   Purchase Address  185950 non-null  object
 6   month             185950 non-null  int32 
dtypes: int32(1), object(6)
memory usage: 10.6+ MB


let's convert Quantity Ordered and Price Each column values to numbers (int and float)

In [37]:
data['Quantity Ordered'] = pd.to_numeric(data['Quantity Ordered'])
data['Price Each'] = pd.to_numeric(data['Price Each'])

### adding sales column to the data
qty * price = sales

In [40]:
data['Sales'] = data['Quantity Ordered'] * data['Price Each']
data.sort_values(by = 'Quantity Ordered', ascending = False)

Order ID                   Product  Quantity Ordered  Price Each  \
163233   227096    AAA Batteries (4-pack)                 9        2.99   
162597   226483    AAA Batteries (4-pack)                 9        2.99   
174464   211097    AAA Batteries (4-pack)                 9        2.99   
160735   224719    AAA Batteries (4-pack)                 8        2.99   
112217   291034    AAA Batteries (4-pack)                 8        2.99   
...         ...                       ...               ...         ...   
65320    171357          Wired Headphones                 1       11.99   
65321    171358  Lightning Charging Cable                 1       14.95   
65322    171359  Lightning Charging Cable                 1       14.95   
65323    171360           Vareebadd Phone                 1      400.00   
186849   222909    AAA Batteries (4-pack)                 1        2.99   

            Order Date                        Purchase Address  month   Sales  
163233  07/29/19 09:50     60 10th St, San Francisco, CA 94016      7   26.91  
162597  07/20/19 23:28  867 Sunset St, San Francisco, CA 94016      7   26.91  
174464  06/23/19 08:29   338 Hickory St, Los Angeles, CA 90001      6   26.91  
160735  07/26/19 23:59   876 Adams St, San Francisco, CA 94016      7   23.92  
112217  11/09/19 16:30         268 Willow St, Dallas, TX 75001     11   23.92  
...                ...                                     ...    ...     ...  
65320   03/03/19 10:42         676 Cedar St, Seattle, WA 98101      3   11.99  
65321   03/07/19 11:28     735 1st St, San Francisco, CA 94016      3   14.95  
65322   03/30/19 19:48   981 River St, San Francisco, CA 94016      3   14.95  
65323   03/26/19 13:10  124 Spruce St, San Francisco, CA 94016      3  400.00  
186849  06/25/19 14:33          209 11th St, Atlanta, GA 30301      6    2.99  

[185950 rows x 8 columns]

## Question 1: What's the best month in terms of sales?

In [59]:
month_wise_sales = data.groupby('month').sum(numeric_only=True)
month_wise_sales

Quantity Ordered  Price Each       Sales
month                                          
1                 10903  1811768.38  1822256.73
2                 13449  2188884.72  2202022.42
3                 17005  2791207.83  2807100.38
4                 20558  3367671.02  3390670.24
5                 18667  3135125.13  3152606.75
6                 15253  2562025.61  2577802.26
7                 16072  2632539.56  2647775.76
8                 13448  2230345.42  2244467.88
9                 13109  2084992.09  2097560.13
10                22703  3715554.83  3736726.88
11                19798  3180600.68  3199603.20
12                28114  4588415.41  4613443.34

In [75]:
import plotly.express as px

fig = px.line(month_wise_sales, x = month_wise_sales.index, y = 'Sales', title = 'Month wise sales')
fig.add_annotation(x = 12, y = 4613443.34, text = 'Peak sales')
fig.update_layout(
    autosize=False,
    width=700,
    height=700)
fig.show()